sktime/forecasting/compose/_networkpipeline.py

```diff
@@ -0,0 +1,278 @@
+    ...            {"fit": {"Z": "original_y"}, "predict": None, "update": None},
+    ...        ),
+    ...        (
+    ...            "forecaster",
+    ...            NaiveForecaster(strategy="mean"),
+    ...            {
+    ...                "fit": {"y": "imputer", "X": "original_X", "fh": "original_fh"},
+    ...                "predict": {"fh": "original_fh"},
+    ...                "update": {"y": "original_y"},
+    ...            },
+    ...        ),
+    ...    ])
+    """
+
+    _required_parameters = ["steps"]
+    _tags = {"univariate-only": True}
+
+    def __init__(self, steps):
+        self.steps = steps
+        self._step_results = {}
+        self._fitted_estimators = {}
+        super().__init__()
+
+    def _iter(self):
+        for name, est, arguments in self.steps:
+            yield name, est, arguments
+
+    def _check_steps_for_values(self, step_name):
+        if step_name in self._step_results:
+            return self._step_results[step_name]
+
+        return None
+
+    def _process_arguments(self, arguments):
+        """
+        Checks arguments for consistency.
+
+        Replace key word 'original' with X
+
+        Replace string step name with step fit return result
+
+        Parameters
+        ----------
+        arguments : dictionary
+            key-value for fit() method of estimator
+        """
+
+        returned_arguments_kwarg = arguments.copy()
+        if arguments is None:
+            return arguments
+        for argument_name, argument_value in returned_arguments_kwarg.items():
+            if argument_value == "original_X":
+                returned_arguments_kwarg[argument_name] = self._X
+                continue
+            if argument_value == "original_y":
+                returned_arguments_kwarg[argument_name] = self._y
+                continue
+            if argument_value == "original_fh":
+                returned_arguments_kwarg[argument_name] = self._fh
+                continue
+
+            if type(argument_value) == list:
+                out = []
+                for list_val in argument_value:
+                    returned_step_value = self._check_steps_for_values(list_val)
+                    if returned_step_value is not None:
+                        out.append(returned_step_value)
+                returned_arguments_kwarg[argument_name] = out
+                continue
+            # go through all steps and look for returned values
+            returned_step_value = self._check_steps_for_values(argument_value)
+            if returned_step_value is not None:
+                returned_arguments_kwarg[argument_name] = returned_step_value
+
+        return returned_arguments_kwarg
+
+    def fit(self, y, X=None, fh=None):
+        self._set_y_X(y, X)
+        self._set_fh(fh)
+
+        for name, est, arguments in self._iter():
+            processed_arguments = {}
+
+            if "fit" in arguments and arguments["fit"] is None:
```

Comment:
Does that mean we may not fit some estimators?

Reply:
The design of the pipeline allows to skip `fit`, `predict` and `update` for each step. I had in mind the use case where you might not want to update the transformers but only the forecaster (last step).

In my prototype for this pipeline I had a step which only had `transform()` without a `fit()`. It was a custom class for compiling a new dataset from extracted features and processed labels in previous steps in the pipeline. The design has changed a little bit since then so we should probably discuss whether it makes sense to have steps that do not implement `fit()`. The line that you highlighted is a legacy of this first prototype.